Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Analytics.hs

```diff
@@ -0,0 +1,593 @@
+{-
+ Copyright 2022-23, Juspay India Pvt Ltd
+
+ This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License
+
+ as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program
+
+ is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+
+ or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. You should have received a copy of
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+
+module SharedLogic.Analytics where
+
+import Data.Time hiding (getCurrentTime)
+import qualified Domain.Types.Person as DP
+import Kernel.Prelude hiding (any, elem, map)
+import qualified Kernel.Storage.Clickhouse.Config as CH
+import Kernel.Storage.Esqueleto.Config (EsqDBReplicaFlow)
+import qualified Kernel.Storage.Hedis as Redis
+import Kernel.Types.Id
+import Kernel.Utils.Common
+import qualified SharedLogic.DriverFlowStatus as SDFStatus
+import qualified Storage.Clickhouse.DailyStats as CDaily
+import qualified Storage.Clickhouse.DriverInformation as CDI
+import qualified Storage.Clickhouse.DriverStats as CDS
+import qualified Storage.Queries.DriverOperatorAssociation as QDOA
+import qualified Storage.Queries.FleetDriverAssociation as QFDA
+import qualified Storage.Queries.FleetOperatorAssociation as QFOA
+import Tools.Error
+
+-- | Find the operator ID for a driver by checking direct association or fleet association
+findOperatorIdForDriver ::
+  ( MonadFlow m,
+    EsqDBFlow m r,
+    CacheFlow m r
+  ) =>
+  Id DP.Person ->
+  m Text
+findOperatorIdForDriver driverId = do
+  -- First try to find direct driver-operator association
+  mbDriverOperatorAssoc <- QDOA.findByDriverId driverId True
+  case mbDriverOperatorAssoc of
+    Just assoc -> pure assoc.operatorId
+    Nothing -> do
+      -- If no direct association, try to find through fleet
+      mbFleetDriverAssoc <- QFDA.findByDriverId driverId True
+      case mbFleetDriverAssoc of
+        Just fleetAssoc -> do
+          -- Found fleet association, now find the operator for this fleet
+          mbFleetOperatorAssoc <- QFOA.findByFleetOwnerIdAndIsActive (Id fleetAssoc.fleetOwnerId) True
+          case mbFleetOperatorAssoc of
+            Just fleetOpAssoc -> pure fleetOpAssoc.operatorId
+            Nothing -> throwError $ InvalidRequest "Driver is associated with fleet but fleet has no operator"
+        Nothing -> throwError $ InvalidRequest "Driver is not associated with any entity"
```

Comment:
Also, don't think we should throw error here, because this was called in `endRideTransaction`, and can affect ride flow
Generally I think that analytics should not throw any errors in main flow